src/java/org/apache/cassandra/security/DefaultSslContextFactoryImpl.java

```diff
@@ -0,0 +1,369 @@
+        {
+            throw new SSLException("Error creating/initializing the SSL Context", e);
+        }
+    }
+
+    @Override
+    public SslContext createNettySslContext(boolean buildTruststore, SocketType socketType, boolean useOpenSsl,
+                                            CipherSuiteFilter cipherFilter) throws SSLException
+    {
+        /*
+            There is a case where the netty/openssl combo might not support using KeyManagerFactory. Specifically,
+            I've seen this with the netty-tcnative dynamic openssl implementation. Using the netty-tcnative
+            static-boringssl works fine with KeyManagerFactory. If we want to support all of the netty-tcnative
+            options, we would need to fall back to passing in a file reference for both a x509 and PKCS#8 private
+            key file in PEM format (see {@link SslContextBuilder#forServer(File, File, String)}). However, we are
+            not supporting that now to keep the config/yaml API simple.
+         */
+        KeyManagerFactory kmf = buildKeyManagerFactory();
+        SslContextBuilder builder;
+        if (socketType == SocketType.SERVER)
+        {
+            builder = SslContextBuilder.forServer(kmf).clientAuth(this.require_client_auth ? ClientAuth.REQUIRE :
+                                                                  ClientAuth.NONE);
+        }
+        else
+        {
+            builder = SslContextBuilder.forClient().keyManager(kmf);
+        }
+
+        builder.sslProvider(useOpenSsl ? SslProvider.OPENSSL : SslProvider.JDK).protocols(getAcceptedProtocols());
+
+        // only set the cipher suites if the opertor has explicity configured values for it; else, use the default
+        // for each ssl implemention (jdk or openssl)
+        if (cipher_suites != null && !cipher_suites.isEmpty())
+            builder.ciphers(cipher_suites, cipherFilter);
+
+        if (buildTruststore)
+            builder.trustManager(buildTrustManagerFactory());
+
+        try
+        {
+            return builder.build();
+        }
+        catch (SSLException e)
+        {
+            throw new SSLException("failed to build the final SslContext object for secure connections", e);
+        }
+    }
+
+    @Override
+    public synchronized void initHotReloading() throws SSLException {
+        boolean hasKeystore = hasKeystore();
+        boolean hasTruststore = hasTruststore();
+
+        if ( hasKeystore || hasTruststore ) {
+            List<HotReloadableFile> fileList = new ArrayList<>();
+            if ( hasKeystore )
+            {
+                fileList.add(new HotReloadableFile(keystore));
+            }
+            if ( hasTruststore ) {
+                fileList.add(new HotReloadableFile(truststore));
+            }
+            hotReloadableFiles = fileList;
+        }
+    }
+
+    @Override
+    public boolean shouldReload()
+    {
+        return hotReloadableFiles.stream().anyMatch(HotReloadableFile::shouldReload);
```

Comment:
what happens when `initHotReloading` populates hotReloadableFiles with one file only? For example, I can remove the truststore file from disk so it would not exist but I would make keystore existing and newer.

So that means that shouldReload would return true because there is a file which should be reloaded, but when reloading is being to be conducted, does not it fail because my other file, truststore, does not exist anymore?

In other words, what happens with reloading when I (e.g. accidentaly) remove the other file?

Reply:
The `initHotReloading` happens during the initialization from the DatabaseDescriptor. Let us say keystore and truststore both files exists at that time. Now question is - what happens when one of those files is removed and `shouldReload` is called? Since Java APIs doesn't throw any errors for [file.lastModified() ](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#lastModified--) in case file doesn't exist, it would think it should reload the file and [log error](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/security/SSLFactory.java#L366) silently and continue. However, based on the current code (with or without this PR) when it actually tries to build the key/trust manager factory it would fail. This is the existing behavior in SSLFactory also.

Reply:
This makes me think if it is worth to change [HotReloadableFile#shouldReload](https://github.com/maulin-vasavada/cassandra/blob/CEP-9/src/java/org/apache/cassandra/security/DefaultSslContextFactory.java#L189) to make sure the new modified time is greater than the previous modified time instead of just `!=` comparison.